readme.md
# Mark of the Afflicted

![Addon Version](https://img.shields.io/github/release/WidgetInteractive/Mark-of-the-Afflicted.svg?maxAge=2592000)  [![License](https://img.shields.io/badge/license-GPL--2.0%2B-red.svg)](https://github.com/WidgetInteractive/Mark-of-the-Afflicted/blob/master/license)

### Welcome to our GitHub Repository

Mark of the Afflicted is a fairly simple addon for Elder Scrolls Online which tracks all of your vampire and werewolf characters and monitors their bite timers. In addition, it provides a handy map system for locating vampire altars and werewolf shrines.

Right now, Mark of the Afflicted doesn't do that terribly much, but I want to take it a lot farther. But, to do that, I need feedback! What can be improved with the existing setup? What would you like to see added? Let me know, and I'll see what I can do!

Feel like supporting? Why not [donate](https://paypal.me/Section214)?

## Installation

1. You can clone the GitHub repository: `https://github.com/WidgetInteractive/Mark-of-the-Afflicted.git`
2. Or download it directly as a ZIP file: `https://github.com/WidgetInteractive/Mark-of-the-Afflicted/archive/master.zip`

This will download the latest developer copy of Mark of the Afflicted.

## Bugs

If you find an issue, let us know [here](https://github.com/WidgetInteractive/Mark-of-the-Afflicted/issues?state=open)!

## Support

This is a developer's portal for Mark of the Afflicted and should _not_ be used for support. Please visit the [support page](http://www.esoui.com/portal.php?uid=31772&a=listbugs) if you need to submit a support request.


changelog
= 1.0.3 (August 17, 2018) =
* Updated: API version

= 1.0.2 (June 1, 2017) =
* Fixed: Minor bugfix

= 1.0.1 (May 30, 2017) =
* Fixed: Minor bugfix

= 1.0.0 (May 28, 2017) =
* Initial release


core/fonts.lua
MotA_Fonts = {}

local NodeFonts = {
	[1] = {name = GetString(MOTA_OPTION_BOLD), font = "BOLD_FONT"},
	[2] = {name = GetString(MOTA_OPTION_MEDIUM), font = "MEDIUM_FONT"},
	[3] = {name = GetString(MOTA_OPTION_CHAT), font = "CHAT_FONT"},
	[4] = {name = GetString(MOTA_OPTION_ANTIQUE), font = "ANTIQUE_FONT"},
	[5] = {name = GetString(MOTA_OPTION_HANDWRITTEN), font = "HANDWRITTEN_FONT"},
	[6] = {name = GetString(MOTA_OPTION_STONE_TABLET), font = "STONE_TABLET_FONT"},
	[7] = {name = GetString(MOTA_OPTION_GAMEPAD_BOLD), font = "GAMEPAD_BOLD_FONT"},
	[8] = {name = GetString(MOTA_OPTION_GAMEPAD_MEDIUM), font = "GAMEPAD_MEDIUM_FONT"},
	[9] = {name = GetString(MOTA_OPTION_ARIAL_NARROW), font = "univers55"},
}

function MotA_Fonts:GetFontChoices()
	local choices = {}

	for k,v in ipairs(NodeFonts) do
		choices[#choices+1] = v.name
	end
	return choices
end

function MotA_Fonts:GetFontByName(name)
	for k,v in ipairs(NodeFonts) do
		if v.name == name then
			return v.font
		end
	end
	return "ZoFontHeader"
end

-- Font outlines
local NodeOutlines = {
	[1] = {name = GetString(MOTA_OPTION_NONE)},
	[2] = {name = GetString(MOTA_OPTION_SOFT_THICK_SHADOW), outline = "soft-shadow-thick"},
	[3] = {name = GetString(MOTA_OPTION_SOFT_THIN_SHADOW), outline = "soft-shadow-thin"},
	[4] = {name = GetString(MOTA_OPTION_SHADOW), outline = "shadow"},
	[5] = {name = GetString(MOTA_OPTION_THICK_OUTLINE), outline = "thick-outline"},
	[6] = {name = GetString(MOTA_OPTION_THIN_OUTLINE), outline = "thin-outline"},
	[7] = {name = GetString(MOTA_OPTION_OUTLINE), outline = "outline"}
}

function MotA_Fonts:GetOutlineChoices()
	local choices = {}

	for k,v in ipairs(NodeOutlines) do
		choices[#choices+1] = v.name
	end
	return choices
end

function MotA_Fonts:GetOutlineByName(outlineName)
	for k,v in ipairs(NodeOutlines) do
		if v.name == outlineName then
			return v.outline
		end
	end
end

-- Font string layout example: "$(BOLD_FONT)|30|soft-shadow-thick"
function MotA_Fonts:Build(font, size, outline)
	local fontString = zo_strformat("$(<<1>>)|<<2>>", font, size)

	if outline then
		fontString = zo_strformat("<<1>>|<<2>>", fontString, outline)
	end

	return fontString
end

function MotA_Fonts:GetTimerLabelFontString(parent)
	local font        = MotA_Fonts:GetFontByName(parent.savedVariables.timers.labelFont)
	local fontOutline = MotA_Fonts:GetOutlineByName(parent.savedVariables.timers.labelOutline)
	local fontSize    = parent.savedVariables.timers.labelSize

	return MotA_Fonts:Build(font, fontSize, fontOutline)
end

function MotA_Fonts:GetTimerTimeFontString(parent)
	local font        = MotA_Fonts:GetFontByName(parent.savedVariables.timers.timeFont)
	local fontOutline = MotA_Fonts:GetOutlineByName(parent.savedVariables.timers.timeOutline)
	local fontSize    = parent.savedVariables.timers.timeSize

	return MotA_Fonts:Build(font, fontSize, fontOutline)
end


core/helpers.lua
MotA_Helpers = {}

-- Check if user is a vampire --------------------------------------------------
function MotA_Helpers:IsVampire()
	local numBuffs       = GetNumBuffs("player")
	local isVampire      = false
	local hasBloodRitual = false
	local readyTime      = -1

	for buffIndex = 1, numBuffs do
		local _, _, endTime, _, _, _, _, _, _, _, abilityId, _ = GetUnitBuffInfo("player", buffIndex)

		if (abilityId == 35771) or (abilityId == 35776) or (abilityId == 35783) or (abilityId == 35792) then
			isVampire = true
		end
	end

	if isVampire then
		local numSkillLines = GetNumSkillLines(SKILL_TYPE_WORLD)

		for skillIndex = 1, numSkillLines do
			local numSkillAbilities = GetNumSkillAbilities(SKILL_TYPE_WORLD, skillIndex)

			for abilityIndex = 1, numSkillAbilities do
				local name, _, _, passive, _, purchased, _ = GetSkillAbilityInfo(SKILL_TYPE_WORLD, skillIndex, abilityIndex)

				if purchased then
					local abilityId = GetSkillAbilityId(SKILL_TYPE_WORLD, skillIndex, abilityIndex, false)

					if abilityId == 33091 then
						hasBloodRitual = true
					end
				end
			end
		end
	end

	if hasBloodRitual then
		for buffIndex = 1, numBuffs do
			local _, startTime, endTime, _, _, _, _, _, _, _, abilityId, _ = GetUnitBuffInfo("player", buffIndex)

			if abilityId == 40359 then
				readyTime = endTime - (GetFrameTimeMilliseconds()/1000) + GetTimeStamp()
			end
		end
	end

	return isVampire, hasBloodRitual, readyTime
end

-- Check vampirism stage -------------------------------------------------------
function MotA_Helpers:GetVampireStage()
	local numBuffs     = GetNumBuffs("player")
	local vampireStage = nil

	for buffIndex = 1, numBuffs do
		local _, _, endTime, _, _, _, _, _, _, _, abilityId, _ = GetUnitBuffInfo("player", buffIndex)


		if (abilityId == 35771) then
			vampireStage = 1
		elseif (abilityId == 35776) then
			vampireStage = 2
		elseif (abilityId == 35783) then
			vampireStage = 3
		elseif (abilityId == 35792) then
			vampireStage = 4
		end
	end

	return vampireStage
end

-- Check if user is a werewolf -------------------------------------------------
function MotA_Helpers:IsWerewolf()
	local numBuffs     = GetNumBuffs("player")
	local isWerewolf   = false
	local hasBloodmoon = false
	local readyTime    = -1

	for buffIndex = 1, numBuffs do
		local _, _, endTime, _, _, _, _, _, _, _, abilityId, _ = GetUnitBuffInfo("player", buffIndex)

		if abilityId == 35658 then
			isWerewolf = true
		end
	end

	if isWerewolf then
		local numSkillLines = GetNumSkillLines(SKILL_TYPE_WORLD)

		for skillIndex = 1, numSkillLines do
			local numSkillAbilities = GetNumSkillAbilities(SKILL_TYPE_WORLD, skillIndex)

			for abilityIndex = 1, numSkillAbilities do
				local name, _, _, passive, _, purchased, _ = GetSkillAbilityInfo(SKILL_TYPE_WORLD, skillIndex, abilityIndex)

				if purchased then
					local abilityId = GetSkillAbilityId(SKILL_TYPE_WORLD, skillIndex, abilityIndex, false)

					if abilityId == 32639 then
						hasBloodmoon = true
					end
				end
			end
		end
	end

	if hasBloodmoon then
		for buffIndex = 1, numBuffs do
			local _, _, endTime, _, _, _, _, _, _, _, abilityId, _ = GetUnitBuffInfo("player", buffIndex)

			if abilityId == 40525 then
				readyTime = endTime - (GetFrameTimeMilliseconds()/1000) + GetTimeStamp()
			end
		end
	end

	return isWerewolf, hasBloodmoon, readyTime
end


lang/strings.lua
local strings = {
	-- Settings menu header
	MOTA_TITLE = "Mark of the Afflicted",

	-- Common use
	MOTA_RELOAD_WARNING = "Changing this setting will reload the UI!",
	MOTA_TIMERS_READY   = "Bite Ready",

	-- Keybindings
	MOTA_KEYBIND_TOGGLE = "Toggle display of timers",

	-- General settings
	MOTA_GENERAL_TITLE           = "General Settings",
	MOTA_ACCOUNT_WIDE            = "Use Account Wide Settings",
	MOTA_ACCOUNT_WIDE_TIP        = "Enable this setting to allow use the same configuration for all of your characters.",
	MOTA_MODULES_HEADER          = "Modules",
	MOTA_MODULES_DESCRIPTION     = "Enabling or disabling modules requires reloading the UI for changes to take effect.",
	MOTA_ENABLE_TIMERS           = "Bite Timers",
	MOTA_ENABLE_TIMERS_TIP       = "Enable this setting to display bite timers.",
	MOTA_RELOAD_UI               = "Reload UI",
	MOTA_RELOAD_UI_TIP           = "Click this button to reload the UI if you enable or disable modules.",

	-- Timers settings
	MOTA_TIMERS_TITLE                       = "Bite Timer Settings",
	MOTA_TIMERS_LABEL_HEADER                = "Label Settings",
	MOTA_TIMERS_BAR_COLORS_HEADER           = "Timer Bar Color Settings",
	MOTA_TIMERS_SHOW_ALL                    = "Show All Timers",
	MOTA_TIMERS_SHOW_ALL_TIP                = "Enable this setting to show timers for all vampire and werewolf characters.",
	MOTA_TIMERS_DISPLAY_HEADER              = "Display Settings",
	MOTA_TIMERS_LABEL_FONT                  = "Label Font",
	MOTA_TIMERS_LABEL_FONT_TIP              = "Set the font used for timer labels.",
	MOTA_TIMERS_LABEL_OUTLINE               = "Label Outline",
	MOTA_TIMERS_LABEL_OUTLINE_TIP           = "Set the outline effect used for timer labels.",
	MOTA_TIMERS_LABEL_SIZE                  = "Label Size",
	MOTA_TIMERS_LABEL_SIZE_TIP              = "Set the font size used for label timers.",
	MOTA_TIMERS_LABEL_COLOR                 = "Label Color",
	MOTA_TIMERS_LABEL_COLOR_TIP             = "Set the color used for timer labels.",
	MOTA_TIMERS_TIME_FONT                   = "Time Font",
	MOTA_TIMERS_TIME_FONT_TIP               = "Set the font used for timer time display.",
	MOTA_TIMERS_TIME_OUTLINE                = "Time Outline",
	MOTA_TIMERS_TIME_OUTLINE_TIP            = "Set the outline effect used for timer time display.",
	MOTA_TIMERS_TIME_SIZE                   = "Time Size",
	MOTA_TIMERS_TIME_SIZE_TIP               = "Set the font size used for timer time display.",
	MOTA_TIMERS_TIME_COLOR                  = "Time Color",
	MOTA_TIMERS_TIME_COLOR_TIP              = "Set the color used for timer time display.",
	MOTA_TIMERS_TIMER_ACTION                = "Timer Action",
	MOTA_TIMERS_TIMER_ACTION_TIP            = "Set whether timers should fill or drain as they progress.",
	MOTA_TIMERS_LABEL_ALIGNMENT             = "Label Alignment",
	MOTA_TIMERS_LABEL_ALIGNMENT_TIP         = "Set whether you prefer timer labels to be right or left aligned.",
	MOTA_TIMERS_WW_BACKGROUND_COLOR         = "Werewolf Timers - Background",
	MOTA_TIMERS_WW_BACKGROUND_COLOR_TIP     = "Set the color used for the background of werewolf timer bars.",
	MOTA_TIMERS_WW_GLOSS_COLOR              = "Werewolf Timers - Gloss",
	MOTA_TIMERS_WW_GLOSS_COLOR_TIP          = "Set the color used for the gloss (highlight) of werewolf timer bars.",
	MOTA_TIMERS_VA_BACKGROUND_COLOR         = "Vampire Timers - Background",
	MOTA_TIMERS_VA_BACKGROUND_COLOR_TIP     = "Set the color used for the background of vampire timer bars.",
	MOTA_TIMERS_VA_GLOSS_COLOR              = "Vampire Timers - Gloss",
	MOTA_TIMERS_VA_GLOSS_COLOR_TIP          = "Set the color used for the gloss (highlight) of vampire timer bars.",
	MOTA_TIMERS_LOCK_UI                     = "Lock UI",
	MOTA_TIMERS_LOCK_UI_TIP                 = "Enable this setting to prevent the timers window from being moved.",
	MOTA_TIMERS_SCALE                       = "UI Scale",
	MOTA_TIMERS_SCALE_TIP                   = "Set the scale ratio for the timer display.",
	MOTA_TIMERS_SPACING                     = "Timer Spacing",
	MOTA_TIMERS_SPACING_TIP                 = "Set the spacing between individual timers.",
	MOTA_TIMERS_HIDE_IN_COMBAT              = "Hide While In Combat",
	MOTA_TIMERS_HIDE_IN_COMBAT_TIP          = "Enable this setting to hide bite timers while in combat.",
	MOTA_TIMERS_NOTIFICATIONS               = "Notification Type",
	MOTA_TIMERS_NOTIFICATIONS_TIP           = "Choose how notifications should be displayed when bites are ready.",
	MOTA_TIMERS_NOTIFICATION_SOUND          = "Notification Sound",
	MOTA_TIMERS_NOTIFICATION_SOUND_TIP      = "Choose the sound effect played when bites are ready.",
	MOTA_TIMERS_NOTIFICATION_SOUND_TEST     = "Test Notification Sound",
	MOTA_TIMERS_NOTIFICATION_SOUND_TEST_TIP = "Click this button to test the currently selected notification sound.",
	MOTA_TIMERS_COMPLETED                   = "Bite ready",

	-- Dropdown option strings
	MOTA_OPTION_NONE                   = "None",
	MOTA_OPTION_CHAT                   = "Chat",
	MOTA_OPTION_ANNOUNCEMENT           = "Announcement",
	MOTA_OPTION_LEFT                   = "Left",
	MOTA_OPTION_RIGHT                  = "Right",
	MOTA_OPTION_BOLD                   = "Bold",
	MOTA_OPTION_MEDIUM                 = "Medium",
	MOTA_OPTION_ANTIQUE                = "Antique",
	MOTA_OPTION_HANDWRITTEN            = "Handwritten",
	MOTA_OPTION_STONE_TABLET           = "Stone Tablet",
	MOTA_OPTION_GAMEPAD_BOLD           = "Gamepad Bold",
	MOTA_OPTION_GAMEPAD_MEDIUM         = "Gamepad Medium",
	MOTA_OPTION_ARIAL_NARROW           = "Arial Narrow",
	MOTA_OPTION_SOFT_THICK_SHADOW      = "Soft Thick Shadow",
	MOTA_OPTION_SOFT_THIN_SHADOW       = "Soft Thin Shadow",
	MOTA_OPTION_SHADOW                 = "Shadow",
	MOTA_OPTION_THICK_OUTLINE          = "Thick Outline",
	MOTA_OPTION_THIN_OUTLINE           = "Thin Outline",
	MOTA_OPTION_OUTLINE                = "Outline",
	MOTA_OPTION_FILL                   = "Fill",
	MOTA_OPTION_DRAIN                  = "Drain"
}
for stringId, stringValue in pairs(strings) do
	ZO_CreateStringId(stringId, stringValue)
	SafeAddVersion(stringId, 1)
end


core/commands.lua
-- Variables -------------------------------------------------------------------
MotA_Slash_Commands = {}

local colorYellow     = "|cFFFF00"
local colorSoftYellow = "|cCCCC00"
local colorRed        = "|cFF0000"
local colorBlue       = "|c1155bb"
local colorWidget     = "|c1fda9a"

-- Initialize slash commands ---------------------------------------------------
function MotA_Slash_Commands:Initialize(parent)
	SLASH_COMMANDS["/mota"] = function(command)
		local options = {}

		for o in command:gmatch("%S+") do
			table.insert(options, o)
		end

		if options[1] then
			local command  = string.lower(options[1])
			local option   = (options[2] and string.lower(options[2]) or nil)
			local argument = (options[3] and options[3] or nil)

			if command == "test" then
				--local uid = GetCurrentCharacterId()
				--d(parent.savedVariables.timers.list[uid])
			else
				MotA_Slash_Commands:DisplayGeneralHelp(parent)
			end
		else
			MotA_Slash_Commands:DisplayGeneralHelp(parent)
		end
	end
end

-- Display general help --------------------------------------------------------
function MotA_Slash_Commands:DisplayGeneralHelp(parent)
	d(colorWidget .. "[Mark of the Afflicted] " .. colorYellow .. GetString(MOTA_COMMANDS_VALID) .. ":")

	d(colorRed .. GetString(MOTA_COMMANDS_NONE))
end


modules/timers.lua
-- Variables -------------------------------------------------------------------
MotA_Timers    = {}
MotA_Timer_Bar = {}

-- Initialize timers -----------------------------------------------------------
function MotA_Timers:Initialize(parent)
	MotA_Timer_Bar:Initialize()

	local uid = GetCurrentCharacterId()

	MotA_Timers.timers        = {}
	MotA_Timers.timerKeys     = {}
	MotA_Timers.parent        = parent

	MotA_BiteTimersContainer:SetAnchor(MotA_Timers.parent.savedVariables.timers.position.point, GuiRoot, MotA_Timers.parent.savedVariables.timers.position.relPoint, MotA_Timers.parent.savedVariables.timers.position.x, MotA_Timers.parent.savedVariables.timers.position.y)
	MotA_BiteTimersContainer:SetScale(MotA_Timers.parent.savedVariables.timers.scale)

	-- Lock the window if lock UI is enabled
	if MotA_Timers.parent.savedVariables.timers.lockUI then
		MotA_BiteTimersContainer:SetMovable(false)
	end

	MotA_BiteTimersContainer:SetHandler( "OnMoveStop", function()
		local _, point, _, relPoint, x, y = MotA_BiteTimersContainer:GetAnchor(0)
		MotA_Timers.parent.savedVariables.timers.position = nil
		MotA_Timers.parent.savedVariables.timers.position = { point = point, relPoint = relPoint, x=x, y=y }
	end )

	if MotA_Timers.parent.savedVariables.enable.timers then
		local isVampire, hasBloodRitual, vampireReadyTime = MotA_Helpers:IsVampire()
		local isWerewolf, hasBloodmoon, werewolfReadyTime = MotA_Helpers:IsWerewolf()

		if MotA_Timers.parent.savedTimers.timers[uid] then
			if not isVampire and not isWerewolf then
				MotA_Timers.parent.savedTimers.timers[uid] = nil
			end
		end

		if isVampire or isWerewolf then
			if isVampire then
				MotA_Timers.parent.savedTimers.timers[uid] = {
					char  = GetUnitName("player"),
					type  = "vampire",
					stage = MotA_Helpers:GetVampireStage(),
					time  = vampireReadyTime
				}
			elseif isWerewolf then
				MotA_Timers.parent.savedTimers.timers[uid] = {
					char  = GetUnitName("player"),
					type  = "werewolf",
					stage = nil,
					time  = werewolfReadyTime
				}
			end
		end

		local timers = MotA_Timers.parent.savedTimers.timers

		table.sort(timers, function(t1, t2)
			if t1[1] < t2[1] then
				return true
			end
			return false
		end)

		for k, v in pairs(timers) do
			if not MotA_Timers.parent.savedVariables.timers.showAll and k == uid then
				MotA_Timers:AddBiteTimer(v["type"], v["char"], v["time"], v["stage"])
			elseif MotA_Timers.parent.savedVariables.timers.showAll then
				MotA_Timers:AddBiteTimer(v["type"], v["char"], v["time"], v["stage"])
			end
		end

		MotA_BiteTimersContainer:RegisterForEvent(EVENT_PLAYER_COMBAT_STATE, function(eventCode, inCombat) MotA_Timers:ShowTimers(inCombat) end)
	end

	local fragment = ZO_SimpleSceneFragment:New(MotA_BiteTimersContainer)
	fragment:RegisterCallback("StateChange", function(oldState, newState)
		if newState == SCENE_FRAGMENT_SHOWING then
			MotA_Timers:UpdateTimers(GetGameTimeMilliseconds()/1000)
		end

		if newState == SCENE_FRAGMENT_SHOWN then
			MotA_BiteTimersContainer:SetHidden(MotA_Timers.parent.savedVariables.timers.isHidden)
		end
	end)

	local scene = SCENE_MANAGER:GetScene("hudui")

	scene:AddFragment(fragment)
	scene = SCENE_MANAGER:GetScene("hud")
	scene:AddFragment(fragment)
end

-- Initialize ------------------------------------------------------------------
function MotA_Timer_Bar:Initialize()
	self.barsPool = ZO_ControlPool:New("MotA_Timer", MotA_BiteTimersContainer, "Bar")
end

-- Add timer -------------------------------------------------------------------
function MotA_Timers:AddBiteTimer(biteType, charName, readyTime, vampireStage)
	local bar = MotA_Timer_Bar:New()
	local key = charName

	bar:CreateNewBar(biteType, charName, readyTime, vampireStage, #self.timerKeys)

	self.timers[key] = bar
	table.insert(self.timerKeys, key)
end

-- Create bar ------------------------------------------------------------------
function MotA_Timer_Bar:CreateNewBar(biteType, charName, readyTime, vampireStage, count)
	self.bar, self.barKey = self.barsPool:AcquireObject()
	self.bar:ClearAnchors()
	self.bar:SetAnchor(TOP, MotA_BiteTimersContainer, TOP, 0, (count*MotA_Timers.parent.savedVariables.timers.spacing)+30)
	MotA_BiteTimersContainer:SetHeight((count+1)*52)

	self.timeLeftLabel = self.bar:GetNamedChild("TimeLeft")
	self.progressBar   = self.bar:GetNamedChild("Bar")
	self.label         = self.bar:GetNamedChild("Label")
	self.label:SetFont(MotA_Fonts:GetTimerLabelFontString(MotA_Timers.parent))
	self.label:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.labelColor))
	self.timeLeftLabel:SetFont(MotA_Fonts:GetTimerTimeFontString(MotA_Timers.parent))
	self.timeLeftLabel:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.timeColor))

	self.biteType      = biteType
	self.uid           = GetCurrentCharacterId()
	self.charName      = charName
	self.timeRemaining = readyTime - GetTimeStamp()
	self.duration      = 604800
	self.vampireStage  = vampireStage

	if biteType == "vampire" then
		self.progressBar:GetNamedChild("Gloss"):SetColor(unpack(MotA_Timers.parent.savedVariables.timers.vaGlossColor))
		self.progressBar:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.vaBackgroundColor))

		self.label:SetText(string.upper(zo_strformat("<<1>> - <<2>> <<3>>", charName, "Stage", vampireStage)))
	else
		self.progressBar:GetNamedChild("Gloss"):SetColor(unpack(MotA_Timers.parent.savedVariables.timers.wwGlossColor))
		self.progressBar:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.wwBackgroundColor))

		self.label:SetText(string.upper(zo_strformat("<<1>>", charName)))
	end

	local now = GetGameTimeMilliseconds()/1000

	self.label:ClearAnchors()

	if MotA_Timers.parent.savedVariables.timers.labelAlignment == GetString(MOTA_OPTION_RIGHT) then
		self.label:SetAnchor(BOTTOMRIGHT, self.progressBar, TOPRIGHT, 0, -4)
		self.progressBar:SetBarAlignment(BAR_ALIGNMENT_REVERSE)
		self.progressBar:GetNamedChild("Gloss"):SetBarAlignment(BAR_ALIGNMENT_REVERSE)
	else
		self.label:SetAnchor(BOTTOMLEFT, self.progressBar, TOPLEFT, 0, -4)
		self.progressBar:SetBarAlignment(BAR_ALIGNMENT_NORMAL)
		self.progressBar:GetNamedChild("Gloss"):SetBarAlignment(BAR_ALIGNMENT_NORMAL)
	end

	self.progressBar:SetMinMax(0, self.duration)

	if MotA_Timers.parent.savedVariables.timers.timerAction == GetString(MOTA_OPTION_DRAIN) then
		self.barValue = self.timeRemaining
	else
		self.barValue = self.duration-self.timeRemaining
	end

	self.updateInterval = self.duration/320
	self.nextBarUpdate  = 0
	self.timeout        = now + self.timeRemaining
	self.nextTimeUpdate = 0
	self.start          = now
	self.bar:SetHandler("OnUpdate", function(control, time) self:Update(time) end)
end

-- Setup new bar ---------------------------------------------------------------
function MotA_Timer_Bar:New()
	local ret    = setmetatable({}, self)
	self.__index = self

	return ret
end

-- Update timers ---------------------------------------------------------------
function MotA_Timers:UpdateTimers(time)
	for i=1, #self.timerKeys do
		self.timers[self.timerKeys[i]]:ForceUpdate(time)
	end
end

-- Show timers ------------------------------------------------------------------
function MotA_Timers:ShowTimers(inCombat)
	if not MotA_Timers.parent.savedVariables.timers.hideInCombat then
		return
	end

	if not MotA_Timers.parent.savedVariables.timers.isHidden then
		MotA_BiteTimersContainer:SetHidden(inCombat)

		if not inCombat then
			self:UpdateTimers(GetGameTimeMilliseconds()/1000)
		end
	end
end

-- Update bar ------------------------------------------------------------------
function MotA_Timer_Bar:Update(time)
	if time > self.timeout then
		self:Completed()
		return
	end

	local _, _, vampireReadyTime = MotA_Helpers:IsVampire()
	local _, _, werewolfReadyTime = MotA_Helpers:IsWerewolf()

	if self.biteType == "vampire" then
		self.vampireStage = MotA_Helpers:GetVampireStage()

		MotA_Timers.parent.savedTimers.timers[self.uid] = {
			char  = GetUnitName("player"),
			type  = "vampire",
			stage = self.vampireStage,
			time  = vampireReadyTime
		}
	else
		MotA_Timers.parent.savedTimers.timers[self.uid] = {
			char  = GetUnitName("player"),
			type  = "werewolf",
			stage = nil,
			time  = werewolfReadyTime
		}
	end

	if self.biteType == "vampire" then
		self.label:SetText(string.upper(zo_strformat("<<1>> - <<2>> <<3>>", self.charName, "Stage", self.vampireStage)))
	end

	if time > self.nextBarUpdate then
		local t = ""

		if MotA_Timers.parent.savedVariables.timers.timerAction == GetString(MOTA_OPTION_DRAIN) then
			t = self.timeRemaining
		else
			t = self.barValue + time - self.start
		end

		self.progressBar:SetValue(t)
		self.nextBarUpdate = time + self.updateInterval
	end

	if time > self.nextTimeUpdate then
		self.timeRemaining = self.timeout - time

		local timerPercent = math.floor(self.timeRemaining/self.duration*100)
		local timerLabel = ""

		timerLabel = ZO_FormatTime(self.timeRemaining, TIME_FORMAT_STYLE_COLONS, TIME_FORMAT_PRECISION_TWELVE_HOUR)

		self.timeLeftLabel:SetText(timerLabel)
		self.nextTimeUpdate = time + 1
	end
end

function MotA_Timer_Bar:ForceUpdate(time)
	if self.completed then
		return
	end

	self.nextBarUpdate  = 0
	self.nextTimeUpdate = 0
	self:Update(time)
end

-- Apply settings on save ------------------------------------------------------
function MotA_Timers:ApplySettings()
	for i=1, #self.timerKeys do
		self.timers[self.timerKeys[i]].label:SetFont(MotA_Fonts:GetTimerLabelFontString(MotA_Timers.parent))
		self.timers[self.timerKeys[i]].label:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.labelColor))
		self.timers[self.timerKeys[i]].timeLeftLabel:SetFont(MotA_Fonts:GetTimerTimeFontString(MotA_Timers.parent))
		self.timers[self.timerKeys[i]].timeLeftLabel:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.timeColor))

		if self.timers[self.timerKeys[i]].biteType == "vampire" then
			self.timers[self.timerKeys[i]].progressBar:GetNamedChild("Gloss"):SetColor(unpack(MotA_Timers.parent.savedVariables.timers.vaGlossColor))
			self.timers[self.timerKeys[i]].progressBar:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.vaBackgroundColor))

			self.timers[self.timerKeys[i]].label:SetText(string.upper(zo_strformat("<<1>> - <<2>> <<3>>", self.timers[self.timerKeys[i]].charName, "Stage", self.timers[self.timerKeys[i]].vampireStage)))
		else
			self.timers[self.timerKeys[i]].progressBar:GetNamedChild("Gloss"):SetColor(unpack(MotA_Timers.parent.savedVariables.timers.wwGlossColor))
			self.timers[self.timerKeys[i]].progressBar:SetColor(unpack(MotA_Timers.parent.savedVariables.timers.wwBackgroundColor))

			self.timers[self.timerKeys[i]].label:SetText(string.upper(zo_strformat("<<1>>", self.timers[self.timerKeys[i]].charName)))
		end
	end

	-- Lock the window if lock UI is enabled
	if MotA_Timers.parent.savedVariables.timers.lockUI then
		MotA_BiteTimersContainer:SetMovable(false)
	else
		MotA_BiteTimersContainer:SetMovable(true)
	end
end

-- Rearrange bars --------------------------------------------------------------
function MotA_Timers:RearrangeBars()
	if #self.timerKeys == 0 then
		MotA_BiteTimersContainer:SetHeight(1)
		return
	end

	local count = 0

	for i=1, #self.timerKeys do
		self.timers[self.timerKeys[i]].bar:ClearAnchors()
		self.timers[self.timerKeys[i]].bar:SetAnchor(TOP, MotA_BiteTimersContainer, TOP, 0, (count*MotA_Timers.parent.savedVariables.timers.spacing)+30)

		MotA_BiteTimersContainer:SetHeight((count+1)*MotA_Timers.parent.savedVariables.timers.spacing)

		self.timers[self.timerKeys[i]].label:ClearAnchors()

		if MotA_Timers.parent.savedVariables.timers.labelAlignment == GetString(MOTA_OPTION_RIGHT) then
			self.timers[self.timerKeys[i]].label:SetAnchor(BOTTOMRIGHT, self.timers[self.timerKeys[i]].progressBar, TOPRIGHT, 0, -4)
			self.timers[self.timerKeys[i]].progressBar:SetBarAlignment(BAR_ALIGNMENT_REVERSE)
			self.timers[self.timerKeys[i]].progressBar:GetNamedChild("Gloss"):SetBarAlignment(BAR_ALIGNMENT_REVERSE)
		else
			self.timers[self.timerKeys[i]].label:SetAnchor(BOTTOMLEFT, self.timers[self.timerKeys[i]].progressBar, TOPLEFT, 0, -4)
			self.timers[self.timerKeys[i]].progressBar:SetBarAlignment(BAR_ALIGNMENT_NORMAL)
			self.timers[self.timerKeys[i]].progressBar:GetNamedChild("Gloss"):SetBarAlignment(BAR_ALIGNMENT_NORMAL)
		end

		count = count + 1
	end
end

-- Complete --------------------------------------------------------------------
function MotA_Timer_Bar:Completed()
	local uid = GetCurrentCharacterId();

	self.bar:SetHandler("OnUpdate", nil)
	self.progressBar:SetMinMax(0, 1)

	if MotA_Timers.parent.savedVariables.timers.timerAction == GetString(MOTA_OPTION_DRAIN) then
		self.progressBar:SetValue(0)
	else
		self.progressBar:SetValue(1)
	end

	self.completed = true

	self.timeLeftLabel:SetText(GetString(MOTA_TIMERS_READY) .. "!")

	if MotA_Timers.parent.savedVariables.timers.notifications == "Chat" then
		PlaySound(MotA_Timers.parent.savedVariables.timers.notificationSound)
		CHAT_SYSTEM:AddMessage(colorWidget .. "[Mark of the Afflicted] " .. colorYellow .. GetString(MOTA_TIMERS_READY) .. ": " .. self.charName)
	elseif MotA_Timers.parent.savedVariables.timers.notifications == "Announcement" then
		CENTER_SCREEN_ANNOUNCE:AddMessage(0, CSA_EVENT_SMALL_TEXT, MotA_Timers.parent.savedVariables.timers.notificationSound, GetString(MOTA_TIMERS_READY) .. self.charName)
	end
end

-- Toggle timers ---------------------------------------------------------------
function MotA_Timers_Toggle()
	MotA_Timers.parent.savedVariables.timers.isHidden = not MotA_Timers.parent.savedVariables.timers.isHidden
	MotA_BiteTimersContainer:SetHidden(MotA_Timers.parent.savedVariables.timers.isHidden)
end
